esphome/components/st7735/st7735.cpp

```diff
@@ -1,474 +1,504 @@
+    0xffe0, 0xffea, 0xfff5, 0xffff};
+
+ST7735::ST7735(ST7735Model model, int width, int height, int colstart, int rowstart, boolean eightbitcolor) {
+  model_ = model;
+  this->width_ = width;
+  this->height_ = height;
+  this->colstart_ = colstart;
+  this->rowstart_ = rowstart;
+  this->eightbitcolor_ = eightbitcolor;
+}
+
+void ST7735::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up ST7735...");
+  this->spi_setup();
+
+  this->dc_pin_->setup();  // OUTPUT
+  this->cs_->setup();      // OUTPUT
+
+  this->dc_pin_->digital_write(true);
+  this->cs_->digital_write(true);
+
+  this->init_reset_();
+  delay(100);  // NOLINT
+
+  ESP_LOGD(TAG, "  START");
+  dump_config();
+  ESP_LOGD(TAG, "  END");
+
+  display_init_(RCMD1);
+
+  if (this->model_ == INITR_GREENTAB) {
+    display_init_(RCMD2GREEN);
+    colstart_ == 0 ? colstart_ = 2 : colstart_;
+    rowstart_ == 0 ? rowstart_ = 1 : rowstart_;
+  } else if ((this->model_ == INITR_144GREENTAB) || (this->model_ == INITR_HALLOWING)) {
+    height_ == 0 ? height_ = ST7735_TFTHEIGHT_128 : height_;
+    width_ == 0 ? width_ = ST7735_TFTWIDTH_128 : width_;
+    display_init_(RCMD2GREEN144);
+    colstart_ == 0 ? colstart_ = 2 : colstart_;
+    rowstart_ == 0 ? rowstart_ = 3 : rowstart_;
+  } else if (this->model_ == INITR_MINI_160X80) {
+    height_ == 0 ? height_ = ST7735_TFTHEIGHT_160 : height_;
+    width_ == 0 ? width_ = ST7735_TFTWIDTH_80 : width_;
+    display_init_(RCMD2GREEN160X80);
+    colstart_ = 24;
+    rowstart_ = 0;  // For default rotation 0
+  } else {
+    // colstart, rowstart left at default '0' values
+    display_init_(RCMD2RED);
+  }
+  display_init_(RCMD3);
+
+  // Black tab, change MADCTL color filter
+  if ((this->model_ == INITR_BLACKTAB) || (this->model_ == INITR_MINI_160X80)) {
+    uint8_t data = 0xC0;
+    sendcommand_(ST77XX_MADCTL, &data, 1);
+  }
+
+  this->init_internal_(this->get_buffer_length());
+  memset(this->buffer_, 0x00, this->get_buffer_length());
+}
+
+void ST7735::update() {
+  this->do_update_();
+  this->write_display_data_();
+}
+
+int ST7735::get_height_internal() { return height_; }
+
+int ST7735::get_width_internal() { return width_; }
+
+size_t ST7735::get_buffer_length() {
+  if (eightbitcolor_) {
+    return size_t(this->get_width_internal()) * size_t(this->get_height_internal());
+  }
+  return size_t(this->get_width_internal()) * size_t(this->get_height_internal()) * 2;
+}
+
+void HOT ST7735::draw_absolute_pixel_internal(int x, int y, Color color) {
+  if (x >= this->get_width_internal() || x < 0 || y >= this->get_height_internal() || y < 0)
+    return;
+
+  if (this->eightbitcolor_) {
+    // 8-Bit color space is in BGR332, and has to be calculated by hand
+    const uint8_t color332 = (scale8(color.blue, 7) << 5) | (scale8(color.green, 7) << 2) | (scale8(color.red, 3) << 0);
+    uint16_t pos = (x + y * this->get_width_internal());
+    this->buffer_[pos++] = color332;
```

Comment:
Should be this->buffer_[pos] = color332;
++ will over run the buffer if someone is using all the columns and rows (like me)

Reply:
I think your color conversion should go in color.h for others to use
uint8_t to_rgb_332() const {
    uint8_t color332 =
        (esp_scale8(this->red, 7) << 5) | (esp_scale8(this->green, 7) << 2) | (esp_scale8(this->blue, 3) << 0);
    return color332;
  }

Want to add it to this PR?

Reply:
If you dont have the time i can pull this all in. Id like to get it in sooner than later since this has yet to be merged. Some folks like to do this on their own but i dont mind.

Reply:
No I don't mind, you may alter it any way you like!

The ++ indeed has no meaning in the current code I think. It only increments the POS after use, but POS is not used anymore after increment, so could safely be removed.
Thanks for the review!

Reply:
yeah true. ill put a PR together. Thanks!

Reply:
Wait! BGR332 and RGB332 are swapped now with the new function in Color.. This will mess up the colors. I will try to fix this..